providers/configcat/src/main/java/dev/openfeature/contrib/providers/configcat/ConfigCatProvider.java

```diff
@@ -0,0 +1,176 @@
+ */
+@Slf4j
+public class ConfigCatProvider extends EventProvider {
+
+    @Getter
+    private static final String NAME = "ConfigCat";
+
+    public static final String PROVIDER_NOT_YET_INITIALIZED = "provider not yet initialized";
+    public static final String UNKNOWN_ERROR = "unknown error";
+
+    private ConfigCatProviderConfig configCatProviderConfig;
+
+    @Getter
+    private ConfigCatClient configCatClient;
+
+    @Getter
+    private ProviderState state = ProviderState.NOT_READY;
+
+    private AtomicBoolean isInitialized = new AtomicBoolean(false);
+
+    /**
+     * Constructor.
+     * @param configCatProviderConfig configCatProvider Config
+     */
+    public ConfigCatProvider(ConfigCatProviderConfig configCatProviderConfig) {
+        this.configCatProviderConfig = configCatProviderConfig;
+    }
+
+    /**
+     * Initialize the provider.
+     * @param evaluationContext evaluation context
+     * @throws Exception on error
+     */
+    @Override
+    public void initialize(EvaluationContext evaluationContext) throws Exception {
+        boolean initialized = isInitialized.getAndSet(true);
+        if (initialized) {
+            throw new GeneralError("already initialized");
+        }
+        super.initialize(evaluationContext);
+        configCatClient = ConfigCatClient.get(configCatProviderConfig.getSdkKey(),
+            configCatProviderConfig.getOptions());
+        configCatProviderConfig.postInit();
+        state = ProviderState.READY;
+        log.info("finished initializing provider, state: {}", state);
+
+        configCatClient.getHooks().addOnConfigChanged(map -> {
+            ProviderEventDetails providerEventDetails = ProviderEventDetails.builder()
+                .flagsChanged(new ArrayList<>(map.keySet()))
+                .message("config changed")
+                .build();
+            emitProviderReady(providerEventDetails);
+        });
+
+        configCatClient.getHooks().addOnError(errorMessage -> {
+            ProviderEventDetails providerEventDetails = ProviderEventDetails.builder()
+                .message(errorMessage)
+                .build();
+            emitProviderError(providerEventDetails);
+        });
+    }
+
+    @Override
+    public Metadata getMetadata() {
+        return () -> NAME;
+    }
+
+    @Override
+    public ProviderEvaluation<Boolean> getBooleanEvaluation(String key, Boolean defaultValue, EvaluationContext ctx) {
+        if (!ProviderState.READY.equals(state)) {
+            if (ProviderState.NOT_READY.equals(state)) {
+                throw new ProviderNotReadyError(PROVIDER_NOT_YET_INITIALIZED);
+            }
+            throw new GeneralError(UNKNOWN_ERROR);
+        }
+        return getEvaluation(Boolean.class, key, defaultValue, ctx);
+    }
+
+    @Override
+    public ProviderEvaluation<String> getStringEvaluation(String key, String defaultValue, EvaluationContext ctx) {
+        if (!ProviderState.READY.equals(state)) {
```

Comment:
These ProviderState checks could be part of the `getEvaluation` method as well to reduce the code duplication.